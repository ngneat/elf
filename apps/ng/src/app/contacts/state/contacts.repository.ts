import { Injectable } from '@angular/core';
import { createStore } from '@ngneat/elf';
import { addEntities, withEntities } from '@ngneat/elf-entities';
import {
  PaginationData,
  selectCurrentPage,
  selectCurrentPageEntities,
  selectPaginationData,
  setCurrentPage,
  setPage,
  updatePaginationData,
  withPagination,
} from '@ngneat/elf-pagination';

interface Contact {
  id: number;
  email: string;
  name: string;
  address: string;
}

const store = createStore(
  { name: 'contacts' },
  withPagination(),
  withEntities<Contact>()
);

@Injectable({ providedIn: 'root' })
export class ContactsRepository {
  activePage$ = store.pipe(selectCurrentPage());
  paginationData$ = store.pipe(selectPaginationData());
  activePageContacts$ = store.pipe(selectCurrentPageEntities());

  setActivePage(id: Contact['id']) {
    store.update(setCurrentPage(id));
  }

  addContacts(page: number, contacts: PaginationData & { data: Contact[] }) {
    const { data, ...pagination } = contacts;

    store.update(
      addEntities(data),
      updatePaginationData(pagination),
      setPage(
        page,
        data.map((c) => c.id)
      )
    );
  }

  get store() {
    return store;
  }
}
