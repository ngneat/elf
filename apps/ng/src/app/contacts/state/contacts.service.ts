import { Injectable } from '@angular/core';
import { getData } from '../contacts-data';
import { timer } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { ContactsRepository } from './contacts.repository';
import { skipWhilePageExists } from '@ngneat/elf-pagination';

@Injectable({ providedIn: 'root' })
export class ContactsService {
  constructor(private repo: ContactsRepository) {}

  getContacts(page: number) {
    return timer(300).pipe(
      map(() => getData({ page })),
      tap((res) => this.repo.addContacts(page, res)),
      skipWhilePageExists(this.repo.store, page)
    );
  }
}
