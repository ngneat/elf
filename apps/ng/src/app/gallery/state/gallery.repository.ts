import { Injectable } from '@angular/core';
import { createStore } from '@ngneat/elf';
import {
  addEntities,
  selectAllEntities,
  selectEntities,
  UIEntitiesRef,
  unionEntities,
  updateEntities,
  withEntities,
  withUIEntities,
} from '@ngneat/elf-entities';

export interface GalleryItem {
  id: number;
  title: string;
  description: string;
  image: string;
}

export interface GalleryUIItem {
  id: number;
  open: boolean;
}

const store = createStore(
  { name: 'gallery' },
  withEntities<GalleryItem>(),
  withUIEntities<GalleryUIItem>()
);

@Injectable({ providedIn: 'root' })
export class GalleryRepository {
  items$ = this.store
    .combine({
      entities: store.pipe(selectAllEntities()),
      UIEntities: store.pipe(selectEntities({ ref: UIEntitiesRef })),
    })
    .pipe(unionEntities());

  addGalleryItems(galleryItems: GalleryItem[]) {
    const uiGalleryItems = galleryItems.map(({ id }) => ({ id, open: false }));
    store.update(
      addEntities(galleryItems),
      addEntities(uiGalleryItems, { ref: UIEntitiesRef })
    );
  }

  get store() {
    return store;
  }

  toggleItemOpen(galleryItemId: number) {
    this.store.update(
      updateEntities(galleryItemId, (item) => ({ ...item, open: !item.open }), {
        ref: UIEntitiesRef,
      })
    );
  }
}
