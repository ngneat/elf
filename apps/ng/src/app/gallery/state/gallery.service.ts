import { Injectable } from '@angular/core';
import { getData } from '../gallery-data';
import { timer } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { GalleryRepository } from './gallery.repository';

@Injectable({ providedIn: 'root' })
export class GalleryService {
  constructor(private repo: GalleryRepository) {}

  getGalleryItems() {
    return timer(300).pipe(
      map(() => getData()),
      tap((res) => this.repo.addGalleryItems(res))
    );
  }
}
