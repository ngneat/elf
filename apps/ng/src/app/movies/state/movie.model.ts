export type Movie = {
  id: number | string;
  title: string;
  genres: (string | number)[];
  actors: (string | number)[];
};

export interface FullMovie {
  id: number | string;
  title: string;
  genres: { id: number | string; name: string }[];
  actors: { id: number | string; name: string }[];
}
