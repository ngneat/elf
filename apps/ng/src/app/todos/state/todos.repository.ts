import { Injectable } from '@angular/core';
import { switchMap } from 'rxjs/operators';
import { write } from '../../store/mutations';
import { createState, select, Store, withProps } from '@ngneat/elf';
import {
  addEntities,
  selectAll,
  selectAllApply,
  updateEntities,
  withEntities,
} from '@ngneat/elf-entities';

interface Todo {
  id: number;
  title: string;
  completed: boolean;
}

export interface TodosProps {
  filter: 'ALL' | 'ACTIVE' | 'COMPLETED';
}

const { state, config } = createState(
  withEntities<Todo>(),
  withProps<TodosProps>({ filter: 'ALL' })
);

const store = new Store({ name: 'todos', state, config });

@Injectable({ providedIn: 'root' })
export class TodosRepository {
  todos$ = store.pipe(selectAll());
  filter$ = store.pipe(select((state) => state.filter));

  visibleTodos$ = this.filter$.pipe(
    switchMap((filter) => {
      return store.pipe(
        selectAllApply({
          filterEntity({ completed }) {
            if (filter === 'ALL') return true;
            return filter === 'COMPLETED' ? completed : !completed;
          },
        })
      );
    })
  );

  addTodo(title: Todo['title']) {
    store.update(addEntities({ id: Math.random(), title, completed: false }));
  }

  updateFilter(filter: TodosProps['filter']) {
    store.update(
      write((state) => {
        state.filter = filter;
      })
    );
  }

  updateCompleted(id: Todo['id']) {
    store.update(
      updateEntities(id, (entity) => ({
        ...entity,
        completed: !entity.completed,
      }))
    );
  }
}
