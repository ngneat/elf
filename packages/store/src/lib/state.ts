type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I
) => void
  ? I
  : never;
type Merge<State extends any[], Key extends PropertyKey> = UnionToIntersection<
  State[number][Key]
>;

export type PropsFactory<Props, Config> = { props: Props; config: Config };
export type EmptyConfig = undefined;

export function createState<S extends PropsFactory<any, any>[]>(
  ...propsFactories: S
): { state: Merge<S, 'props'>; config: Merge<S, 'config'> } {
  const result = { config: {}, state: {} };

  for (const { config, props } of propsFactories) {
    Object.assign(result.config, config);
    Object.assign(result.state, props);
  }

  return result as { state: Merge<S, 'props'>; config: Merge<S, 'config'> };
}
