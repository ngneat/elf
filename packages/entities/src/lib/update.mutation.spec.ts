import {
  createEntitiesStore,
  createTodo,
  createUIEntityStore,
  createUITodo,
  Todo,
  toMatchSnapshot,
} from '@ngneat/elf-mocks';
import { addEntities } from './add.mutation';
import { UIEntitiesRef } from './entity.state';
import {
  updateAllEntities,
  updateEntities,
  updateEntitiesByPredicate,
  updateEntitiesIds,
  upsertEntities,
  upsertEntitiesById,
} from './update.mutation';

describe('update', () => {
  let store: ReturnType<typeof createEntitiesStore>;

  beforeEach(() => {
    store = createEntitiesStore();
  });

  it('should update one entity', () => {
    store.update(addEntities(createTodo(1)));
    toMatchSnapshot(expect, store, 'completed false');
    store.update(updateEntities(1, { completed: true }));
    toMatchSnapshot(expect, store, 'completed true');
  });

  it('should update multiple entities', () => {
    store.update(addEntities([createTodo(1), createTodo(2)]));
    toMatchSnapshot(expect, store, 'multi completed false');
    store.update(updateEntities([1, 2], { completed: true }));
    toMatchSnapshot(expect, store, 'multi completed true');
  });

  it('should update by callback', () => {
    store.update(addEntities(createTodo(1)));
    toMatchSnapshot(expect, store, 'completed false');
    store.update(
      updateEntities(1, (todo) => ({ ...todo, completed: !todo.completed }))
    );
    toMatchSnapshot(expect, store, 'completed true');
  });

  it('should update by predicate', () => {
    store.update(addEntities([createTodo(1), createTodo(2)]));
    toMatchSnapshot(expect, store, 'completed false');
    store.update(
      updateEntitiesByPredicate((entity) => entity.id === 1, {
        completed: true,
      })
    );
    toMatchSnapshot(expect, store, 'completed true');
  });

  it('should update all', () => {
    store.update(addEntities([createTodo(1), createTodo(2)]));
    toMatchSnapshot(expect, store, 'completed false');
    store.update(updateAllEntities({ completed: true }));
    toMatchSnapshot(expect, store, 'completed true');
  });

  it('should work with ref', () => {
    const store = createUIEntityStore();
    store.update(
      addEntities([createUITodo(1), createUITodo(2)], { ref: UIEntitiesRef })
    );
    toMatchSnapshot(expect, store, 'open false');
    store.update(updateEntities(1, { open: true }, { ref: UIEntitiesRef }));
    toMatchSnapshot(expect, store, 'open true');
  });

  it('should work with ref update all', () => {
    const store = createUIEntityStore();
    store.update(
      addEntities([createUITodo(1), createUITodo(2)], { ref: UIEntitiesRef })
    );
    toMatchSnapshot(expect, store, 'open false');
    store.update(updateAllEntities({ open: true }, { ref: UIEntitiesRef }));
    toMatchSnapshot(expect, store, 'open true');
  });

  describe('UpsertById', () => {
    const updater = (e: Todo) => ({ ...e, title: 'elf' });
    const creator = (id: number) => createTodo(id);
    const options = {
      updater,
      creator,
    };

    it(`should add the entity if it doesn't exists`, () => {
      const store = createEntitiesStore();
      store.update(
        addEntities([createTodo(1)]),
        upsertEntitiesById(2, options)
      );
      toMatchSnapshot(expect, store, 'two entities');
    });

    it('should update an entity if exists', () => {
      const store = createEntitiesStore();
      store.update(
        addEntities([createTodo(1)]),
        upsertEntitiesById(1, options)
      );
      toMatchSnapshot(expect, store, 'one entity, title "elf"');
    });

    it('should add the missing entities and update existing', () => {
      const store = createEntitiesStore();
      store.update(
        addEntities([createTodo(1)]),
        upsertEntitiesById([1, 2], options)
      );
      toMatchSnapshot(
        expect,
        store,
        'two entities, 1: title "elf", 2: title "todo 2"'
      );
    });

    it('should merge updater with creator', () => {
      const store = createEntitiesStore();
      store.update(
        addEntities([createTodo(1)]),
        upsertEntitiesById([1, 2], {
          ...options,
          mergeUpdaterWithCreator: true,
        })
      );
      toMatchSnapshot(expect, store, 'two entities, title "elf"');
    });

    it('should work with ref', () => {
      const store = createUIEntityStore();
      store.update(
        addEntities([createUITodo(1)], { ref: UIEntitiesRef }),
        upsertEntitiesById([1, 2], {
          updater: (e) => ({ ...e, open: false }),
          creator: (id) => createUITodo(id),
          mergeUpdaterWithCreator: true,
          ref: UIEntitiesRef,
        })
      );
      toMatchSnapshot(expect, store, 'two entities, open true');
    });
  });

  describe('Upsert', () => {
    it(`should add the entity if it doesn't exists`, () => {
      const store = createEntitiesStore();
      store.update(upsertEntities([createTodo(1)]));
      toMatchSnapshot(expect, store, 'one entities');
    });

    it(`should update the entity if it has the same id`, () => {
      const store = createEntitiesStore();

      const todo = createTodo(1);
      store.update(addEntities([todo]));

      // update the todo
      store.update(upsertEntities([{ id: 1, completed: true }]));

      toMatchSnapshot(expect, store, 'updated entity with completed: true');
    });

    it('should work with ref', () => {
      const store = createUIEntityStore();
      store.update(
        addEntities([createUITodo(1)], { ref: UIEntitiesRef }),
        upsertEntities([{ id: 1, open: true }], { ref: UIEntitiesRef })
      );
      toMatchSnapshot(expect, store, 'one ui entity, open true');
    });

    it('if no ids are added, the id array should be the same ref', () => {
      const store = createEntitiesStore();
      store.update(addEntities([createTodo(1), createTodo(2)]));
      const ids = store.getValue().ids;

      store.update(upsertEntities([{ id: 1, completed: true }]));
      const sameIds = store.getValue().ids;

      expect(ids).toBe(sameIds);
    });
  });

  describe('UpdateEntitiesIds', () => {
    it('should support id update', () => {
      store.update(addEntities([createTodo(1)]));
      toMatchSnapshot(expect, store, 'id updated false');
      store.update(updateEntitiesIds(1, 2));
      toMatchSnapshot(expect, store, 'id updated true');
    });

    it('should update multiple ids', () => {
      store.update(addEntities([createTodo(1), createTodo(2), createTodo(3)]));
      toMatchSnapshot(expect, store, 'ids updated false');
      store.update(updateEntitiesIds([2, 3], [4, 5]));
      toMatchSnapshot(expect, store, 'ids updated true');
    });

    it('should throw if new id already exists in the store', () => {
      store.update(addEntities([createTodo(1), createTodo(2)]));
      expect(() => {
        store.update(updateEntitiesIds(1, 2));
      }).toThrow();
    });

    it('should work with ref', () => {
      const store = createUIEntityStore();
      store.update(addEntities([createUITodo(1)], { ref: UIEntitiesRef }));
      toMatchSnapshot(expect, store, 'id updated false');
      store.update(updateEntitiesIds(1, 2, { ref: UIEntitiesRef }));
      toMatchSnapshot(expect, store, 'id updated true');
    });

    it('should update entity after changing the id', () => {
      store.update(addEntities([createTodo(1)]));
      store.update(
        updateEntitiesIds(1, 2),
        updateEntities(2, { completed: true })
      );
      toMatchSnapshot(expect, store, 'id updated true, completed true');
    });
  });
});
