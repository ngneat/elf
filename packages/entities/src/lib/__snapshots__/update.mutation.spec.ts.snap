// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update UpdateEntitiesIds should support id update: id updated false 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
  },
  "ids": Array [
    1,
  ],
}
`;

exports[`update UpdateEntitiesIds should support id update: id updated true 1`] = `
Object {
  "entities": Object {
    "2": Object {
      "completed": false,
      "id": 2,
      "title": "todo 1",
    },
  },
  "ids": Array [
    2,
  ],
}
`;

exports[`update UpdateEntitiesIds should update entity after changing the id: id updated true, completed true 1`] = `
Object {
  "entities": Object {
    "2": Object {
      "completed": true,
      "id": 2,
      "title": "todo 1",
    },
  },
  "ids": Array [
    2,
  ],
}
`;

exports[`update UpdateEntitiesIds should update multiple ids: ids updated false 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
    "2": Object {
      "completed": false,
      "id": 2,
      "title": "todo 2",
    },
    "3": Object {
      "completed": false,
      "id": 3,
      "title": "todo 3",
    },
  },
  "ids": Array [
    1,
    2,
    3,
  ],
}
`;

exports[`update UpdateEntitiesIds should update multiple ids: ids updated true 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
    "4": Object {
      "completed": false,
      "id": 4,
      "title": "todo 2",
    },
    "5": Object {
      "completed": false,
      "id": 5,
      "title": "todo 3",
    },
  },
  "ids": Array [
    1,
    4,
    5,
  ],
}
`;

exports[`update UpdateEntitiesIds should work with ref: id updated false 1`] = `
Object {
  "UIEntities": Object {
    "1": Object {
      "id": 1,
      "open": false,
    },
  },
  "UIIds": Array [
    1,
  ],
}
`;

exports[`update UpdateEntitiesIds should work with ref: id updated true 1`] = `
Object {
  "UIEntities": Object {
    "2": Object {
      "id": 2,
      "open": false,
    },
  },
  "UIIds": Array [
    2,
  ],
}
`;

exports[`update Upsert should add the entity if it doesn't exists: one entities 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
  },
  "ids": Array [
    1,
  ],
}
`;

exports[`update Upsert should update the entity if it has the same id: updated entity with completed: true 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": true,
      "id": 1,
      "title": "todo 1",
    },
  },
  "ids": Array [
    1,
  ],
}
`;

exports[`update Upsert should work with ref: one ui entity, open true 1`] = `
Object {
  "UIEntities": Object {
    "1": Object {
      "id": 1,
      "open": true,
    },
  },
  "UIIds": Array [
    1,
  ],
}
`;

exports[`update UpsertById should add the entity if it doesn't exists: two entities 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
    "2": Object {
      "completed": false,
      "id": 2,
      "title": "todo 2",
    },
  },
  "ids": Array [
    1,
    2,
  ],
}
`;

exports[`update UpsertById should add the missing entities and update existing: two entities, 1: title "elf", 2: title "todo 2" 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "elf",
    },
    "2": Object {
      "completed": false,
      "id": 2,
      "title": "todo 2",
    },
  },
  "ids": Array [
    1,
    2,
  ],
}
`;

exports[`update UpsertById should merge updater with creator: two entities, title "elf" 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "elf",
    },
    "2": Object {
      "completed": false,
      "id": 2,
      "title": "elf",
    },
  },
  "ids": Array [
    1,
    2,
  ],
}
`;

exports[`update UpsertById should update an entity if exists: one entity, title "elf" 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "elf",
    },
  },
  "ids": Array [
    1,
  ],
}
`;

exports[`update UpsertById should work with ref: two entities, open true 1`] = `
Object {
  "UIEntities": Object {
    "1": Object {
      "id": 1,
      "open": false,
    },
    "2": Object {
      "id": 2,
      "open": false,
    },
  },
  "UIIds": Array [
    1,
    2,
  ],
}
`;

exports[`update should update all: completed false 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
    "2": Object {
      "completed": false,
      "id": 2,
      "title": "todo 2",
    },
  },
  "ids": Array [
    1,
    2,
  ],
}
`;

exports[`update should update all: completed true 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": true,
      "id": 1,
      "title": "todo 1",
    },
    "2": Object {
      "completed": true,
      "id": 2,
      "title": "todo 2",
    },
  },
  "ids": Array [
    1,
    2,
  ],
}
`;

exports[`update should update by callback: completed false 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
  },
  "ids": Array [
    1,
  ],
}
`;

exports[`update should update by callback: completed true 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": true,
      "id": 1,
      "title": "todo 1",
    },
  },
  "ids": Array [
    1,
  ],
}
`;

exports[`update should update by predicate: completed false 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
    "2": Object {
      "completed": false,
      "id": 2,
      "title": "todo 2",
    },
  },
  "ids": Array [
    1,
    2,
  ],
}
`;

exports[`update should update by predicate: completed true 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": true,
      "id": 1,
      "title": "todo 1",
    },
    "2": Object {
      "completed": false,
      "id": 2,
      "title": "todo 2",
    },
  },
  "ids": Array [
    1,
    2,
  ],
}
`;

exports[`update should update multiple entities: multi completed false 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
    "2": Object {
      "completed": false,
      "id": 2,
      "title": "todo 2",
    },
  },
  "ids": Array [
    1,
    2,
  ],
}
`;

exports[`update should update multiple entities: multi completed true 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": true,
      "id": 1,
      "title": "todo 1",
    },
    "2": Object {
      "completed": true,
      "id": 2,
      "title": "todo 2",
    },
  },
  "ids": Array [
    1,
    2,
  ],
}
`;

exports[`update should update one entity: completed false 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
  },
  "ids": Array [
    1,
  ],
}
`;

exports[`update should update one entity: completed true 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": true,
      "id": 1,
      "title": "todo 1",
    },
  },
  "ids": Array [
    1,
  ],
}
`;

exports[`update should work with ref update all: open false 1`] = `
Object {
  "UIEntities": Object {
    "1": Object {
      "id": 1,
      "open": false,
    },
    "2": Object {
      "id": 2,
      "open": false,
    },
  },
  "UIIds": Array [
    1,
    2,
  ],
}
`;

exports[`update should work with ref update all: open true 1`] = `
Object {
  "UIEntities": Object {
    "1": Object {
      "id": 1,
      "open": true,
    },
    "2": Object {
      "id": 2,
      "open": true,
    },
  },
  "UIIds": Array [
    1,
    2,
  ],
}
`;

exports[`update should work with ref: open false 1`] = `
Object {
  "UIEntities": Object {
    "1": Object {
      "id": 1,
      "open": false,
    },
    "2": Object {
      "id": 2,
      "open": false,
    },
  },
  "UIIds": Array [
    1,
    2,
  ],
}
`;

exports[`update should work with ref: open true 1`] = `
Object {
  "UIEntities": Object {
    "1": Object {
      "id": 1,
      "open": true,
    },
    "2": Object {
      "id": 2,
      "open": false,
    },
  },
  "UIIds": Array [
    1,
    2,
  ],
}
`;
